### PR TITLE
preserve session id between requests

### PR DESCRIPTION
**Pull request by volodymyrss**
this should allow to preserve session id between requests. This is what frontend does, and it allows not to repeat post-processing if it's already done.

### CODE REVIEW

**Comment by volodymyrss** [comment]
@burnout87  this is tightly related to dispatcher, so it's good if you are aware.

**Comment by volodymyrss** [comment]
does it need to be returned? You mean, for cross-check? That could be useful.
I thought it is not used if returned?

File: oda_api/api.py [hunk 1]
```diff
@@ -205,10 +211,17 @@ def build_from_envs(cls):
 
         return cls(host=host_url, instrument='mock', cookies=cookies, protocol='http')
 
-    def generate_session_id(self,size=16):
+    def generate_session_id(self, size=16):
```
[the remark this answers is not in the document]
Comment (volodymyrss) [reply oda_api/api.py:214]: sorry I did not manage to understand what you mean here almost at all.
Could you please clarify?

[the remark this answers is not in the document]
Comment (burnout87) [reply oda_api/api.py:214]: Unless I am confusing the session_id, in the dispatcher there's already a method to generate a session_id, used also the creation of the scratch directory for the logs. I saw that this function also creates a session_id, so I was wondering if it is planned to use within the dispatcher, unless I am confusing the scope

**Comment by volodymyrss** [comment]
but, as far as this PR is concerned, we will continue to send session_id to be compatible with older API versions. Right?

**Comment by burnout87** [comment]
> but, as far as this PR is concerned, we will continue to send session_id to be compatible with older API versions. Right?

What do you mean with sending, as far as I have understood so far, the session id is not sent, I was just trying to provide it also in the output meta data

**Comment by volodymyrss** [comment]
I mean oda_api sends the session_id to the dispatcher. Not dispatcher back (which is used only when session_id==new and really this does not happen anymore ever, I think)

it's set [here](https://github.com/oda-hub/oda_api/blob/master/oda_api/api.py#L633) and later [sent](https://github.com/oda-hub/oda_api/blob/master/oda_api/api.py#L449).

**Comment by burnout87** [comment]
It's good to maintain the retro compatibility for sure.In which point is the session_id  received by the dispatcher? I am trying to figure this out and better understand this part

**Comment by volodymyrss** [comment]
the principle of session derives from frontend, drupal in this case. frontend environment has session as long as it's open in browser. this environment can only trigger one action at a time. as you might have seen, it is kind of single-threaded: if the request is sent, the rest is blocked.
the environment also keeps some properties, variables. I.e.  user can set some things which will be transmited in the following request with this session. New session starts fresh.
session is used by dispatcher to distinguish different request sequences.
request sequence consists of requests to backends and post-processing operations
you may notice scratch directories contain session and jobid. 
jobid is derived from parameter selection. i.e. identical requests will have identical jobid.  this allows to re-use results from the same job.
but they will have different session_id. which allows to distinguish requests and describe them according to the request origin (session id) and parameters (job_id)

**Comment by burnout87** [comment]
Thanks for the explanation, it makes it more clear for me. Are you planning  in building some test for this purpose?

**Comment by volodymyrss** [comment]
oh, maybe  we should, it's easier  than explaining. but really in the future.

**Comment by volodymyrss** [comment]
or you mean just this particular PR here?

**Comment by burnout87** [comment]
No, I am talking in general, some tests related to the context of this PR will help in understanding more how a session_id is preserved, as well as verify if it actually works.

**Comment by volodymyrss** [comment]
I will! but in the future.

https://github.com/oda-hub/oda_api/issues/11

**Review by burnout87 — APPROVED** [review]

**Pull request merged**